### PR TITLE
Shared variable and misc linting changes

### PR DESCRIPTION
**Pull request by robert-wagner**
This pr adds the shared variable class and various linting changes to the torch code. It still needs testing for compatibility with the hooking code and unit tests written to verify accuracy. A demo is forthcoming after testing with the hooking code

### CODE REVIEW

**Comment by robert-wagner** [comment]
Depends on the #176

**Comment by iamtrask** [comment]
Merged #176 although it looks like there might be some merge conflicts now.

**Comment by robert-wagner** [comment]
There are. I'll fix them in 5ish hours. Have a quiz

File: grid/lib/SharedVariable.py [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+    return [first, second]
+
+
+def reconstruct(shares):
+    return sum(shares) % field
+
+
+def swap_shares(share, party, interface):
+    if (party == 0):
+        interface.send(share)
```
[the remark this answers is not in the document]
Comment (jvmancuso) [reply grid/lib/SharedVariable.py]: Should probably push that to the next PR though

File: grid/services/torch/hook_service.py [hunk 2]
```diff
@@ -462,34 +453,33 @@ def hook_tensor(self, tensor_type):
         self.hook_get_(tensor_type)
         tu.hook_tensor__ser(self, tensor_type)
 
-
-    def hook_variable(self):
```
Comment (jvmancuso) [inline grid/services/torch/hook_service.py]: Please return to the way it was before.  If we hook over torch.nn.Parameter or SharedVariable, we'll be double-hooking.

File: grid/lib/SharedVariable.py [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+    @staticmethod
+    def forward(ctx, a, b):
+        return spdz_add(a, b)
+        # compute a + b on Shared data - they are regular PyTorch tensors
+
+    @staticmethod
+    def backward(ctx, grad_out):
+        grad_out = grad_out.data
+        return grad_out, grad_out
+        # not grad_out operators are overloaded
```
[the remark this answers is not in the document]
Comment (robert-wagner) [reply grid/lib/SharedVariable.py]: Idk patszke's comment in code I started with. Ill delete the whole comment

File: grid/lib/SharedVariable.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (robert-wagner) [reply grid/lib/SharedVariable.py]: So we need to just pass an interface for sending. Was planning this for the next pr

File: grid/lib/SharedVariable.py [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+    def matmul(self, other):
+        return SharedVariable(SharedMatmul.apply(self.var, other.var, self.party, self.interface), self.party, self.interface, self.requires_grad)
+
+    def grad(self):
+        return self.var.grad
+
+    def t_(self):
+        self.var = self.var.t_()
+
+    def __repr__(self):
```
[the remark this answers is not in the document]
Comment (robert-wagner) [reply grid/lib/SharedVariable.py]: We could hook it the same way as var

File: grid/services/torch/hook_service.py (hunk 2 above)
Comment (robert-wagner) [reply grid/services/torch/hook_service.py]: Which way? I just accepted your changes. My apologies if I messed that up

File: grid/lib/SharedVariable.py [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+    def add(self, other):
+        return SharedVariable(SharedAdd.apply(self.var, other.var), self.party, self.interface, self.requires_grad)
+
+    def mul(self, other):
+        return SharedVariable(SharedMult.apply(self.var, other.var, self.party, self.interface), self.party, self.interface, self.requires_grad)
+
+    def matmul(self, other):
+        return SharedVariable(SharedMatmul.apply(self.var, other.var, self.party, self.interface), self.party, self.interface, self.requires_grad)
+
+    def grad(self):
```
[the remark this answers is not in the document]
Comment (robert-wagner) [reply grid/lib/SharedVariable.py]: For grad?

File: grid/lib/SharedVariable.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jvmancuso) [reply grid/lib/SharedVariable.py]: I think we can assume that certain torch methods will be identical across backends, e.g. `send_` and `get_` would likely be the ones you need here and would be generalized.  Esp. since the actual IPFS sending in Grid will be abstracted from the JSON creation in PySyft.

[the remark this answers is not in the document]
Comment (jvmancuso) [reply grid/lib/SharedVariable.py]: So here that would mean share.send_(worker) and share.get_().

File: grid/lib/SharedVariable.py [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+    @staticmethod
+    def backward(ctx, grad_out, party, interface):
+        a = ctx.saved_tensors
+        ones = encode(torch.FloatTensor(a.shape).one_())
+        return spdz_mul(a, public_add(ones, -a, party), party, interface)
+
+
+class SharedVariable(object):
+
+    def __init__(self, var, party, interface, requires_grad=True):
```
[the remark this answers is not in the document]
Comment (jvmancuso) [reply grid/lib/SharedVariable.py]: Let's wait until we test autograd across IPFS, but I'd tentatively suggest requiring input be a Variable.

File: grid/lib/SharedVariable.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (robert-wagner) [reply grid/lib/SharedVariable.py]: So we would be hooking pytorch for everything? I was planning on having this code work without having pytorch hooked. This would work by having an interface class which had send and recieve methods. There would be a worker interface and a client interface. They would both wrap the grid methods to provide that interface

File: grid/lib/SharedVariable.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (robert-wagner) [reply grid/lib/SharedVariable.py]: That works for me. That also will be easier with pytorch 4

File: grid/lib/SharedVariable.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jvmancuso) [reply grid/lib/SharedVariable.py]: We'll work through this later.

File: grid/services/torch/hook_service.py (hunk 2 above)
Comment (jvmancuso) [reply grid/services/torch/hook_service.py]: No worries, that may have been from an intermediate state.  `hook_variable` shouldn't accept a variable type argument, so any code that calls that function will need to be modified.

File: grid/lib/SharedVariable.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jvmancuso) [reply grid/lib/SharedVariable.py]: Yes

File: grid/lib/SharedVariable.py [hunk 7]
```diff
@@ -0,0 +1,338 @@
...
+        return self.var.grad
+
+    def t_(self):
+        self.var = self.var.t_()
+
+    def __repr__(self):
+        return self.var.__repr__()
+
+    def type(self):
+        return 'SharedVariable'
```
[the remark this answers is not in the document]
Comment (robert-wagner) [reply grid/lib/SharedVariable.py]: For data property do we want it to return var or var.data

[the remark this answers is not in the document]
Comment (jvmancuso) [reply grid/lib/SharedVariable.py]: The latter

[the remark this answers is not in the document]
Comment (jvmancuso) [reply grid/lib/SharedVariable.py]: The latter

File: grid/lib/shared_variable.py [hunk 8]
```diff
@@ -0,0 +1,339 @@
...
+    @staticmethod
+    def backward(ctx, grad_out, party, interface):
+        a = ctx.saved_tensors
+        ones = encode(torch.FloatTensor(a.shape).one_())
+        return spdz_mul(a, public_add(ones, -a, party), party, interface)
+
+
+class SharedVariable(object):
+
+    def __init__(self, var, party, interface, requires_grad=True):
```
Comment (jvmancuso) [inline grid/lib/shared_variable.py]: Can you add a `volatile` argument here?

File: grid/services/torch/hook_worker_service.py [hunk 9]
```diff
@@ -91,32 +90,31 @@ def compile_result(self, result, owners):
...
+                                        id=obj_id, owners=[new_owner])
             response_str = obj._ser()
         else:
             # TODO: replace this with something that triggers a nicer
             #       error on the client
             response_str = 'n/a - tensor not found'
 
-        self.worker.publish(channel=response_channel, message=response_str)
\ No newline at end of file
+        self.worker.publish(channel=response_channel, message=response_str)
```
Comment (jvmancuso) [inline grid/services/torch/hook_worker_service.py]: Please add newlines to end of each file here.

File: grid/lib/shared_variable.py (hunk 8 above)
Comment (robert-wagner) [reply grid/lib/shared_variable.py]: Rather than adding volatile argument made it only accept a variable. This will work well with switch to pytorch 4
